### PR TITLE
[BOO] allow `BooConv2d` to take advantage of non-contiguous memory formats

### PR DESCRIPTION
**Pull request by zjgarvey**
In particular, this should allow setting `model = model.to(memory_format=torch.channels_last)` to enable `NHWC` convolutions, whereas previously we had been forcing these tensors back to `NHWC` format by calling `contiguous()` on the inputs inside the `Launchable` call. 

Tests are added to check if `BooConv2d` is calling into the expected kernel signatures.

### CODE REVIEW

File: iree/turbine/kernel/boo/modeling/replace.py [hunk 1]
```diff
@@ -29,34 +29,52 @@ def __init__(
...
             self.bias = torch.nn.Parameter(torch.randn(out_channels))
         else:
             self.register_parameter("bias", None)
 
     def forward(self, x):
+        input_layout = "NCHW"
+        kernel_layout = "NCHW"
+        output_layout = "NCHW"
+        if x.is_contiguous(memory_format=torch.channels_last):
+            x = x.permute([0, 2, 3, 1])
```
Comment (rkayaith) [inline iree/turbine/kernel/boo/modeling/replace.py:54]: is there any case where the input may not be 4D? e.g. does torch's conv2d support a variable number of batch dims?

File: iree/turbine/kernel/boo/modeling/replace.py [hunk 2]
```diff
@@ -29,34 +29,52 @@ def __init__(
...
+            input_layout = "NHWC"
+        w = self.weight
+        if w.is_contiguous(memory_format=torch.channels_last):
+            w = w.permute([0, 2, 3, 1])
+            kernel_layout = "NHWC"
+            output_layout = "NHWC"
```
Comment (rkayaith) [inline iree/turbine/kernel/boo/modeling/replace.py:60]: is there any particular reason to match the output layout to the weights layout rather than the input layout?

**Review by rkayaith — APPROVED** [review]
nice, LGTM! just some minor questions

File: iree/turbine/kernel/boo/modeling/replace.py (hunk 1 above)
Comment (zjgarvey) [reply iree/turbine/kernel/boo/modeling/replace.py:54]: That's a good point. I'll look into adding support for this case.

Comment (zjgarvey) [reply iree/turbine/kernel/boo/modeling/replace.py:54]: Their conv2d allows rank 3 or 4 input tensors and returns a tensor of the same rank.

File: iree/turbine/kernel/boo/modeling/replace.py (hunk 2 above)
Comment (zjgarvey) [reply iree/turbine/kernel/boo/modeling/replace.py:60]: Yeah, my reasoning for this is:

It's possible the initial model inputs aren't in `channels_last` format. If the weights are in `channels_last` format, I'm expecting that the rest of the model is going to be using `channels_last`, and we should try to produce outputs that are also `channels_last` to speed-up subsequent convolutions.

File: iree/turbine/kernel/boo/modeling/replace.py [hunk 3]
```diff
@@ -29,34 +29,55 @@ def __init__(
         super(BooConv2d, self).__init__()
         self.in_channels = in_channels
         self.out_channels = out_channels
-        self.kernel_size = kernel_size
+        self.kernel_size = (
```
Comment (MaheshRavishankar) [inline iree/turbine/kernel/boo/modeling/replace.py:32]: I dont think I follow this change?

**Review by MaheshRavishankar — COMMENTED** [review]
Just have one question otherwise it looks good to me.

File: iree/turbine/kernel/boo/modeling/replace.py (hunk 3 above)
Comment (zjgarvey) [reply iree/turbine/kernel/boo/modeling/replace.py:32]: @MaheshRavishankar Does the above comment address your concern? 

I apologize for merging without hearing back from you on this comment.
